File: ActionUIAndroid/library/src/main/java/com/abracode/actionui/Views/LabeledContent.kt
package com.abracode.actionui.Views

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.width
import androidx.compose.material3.Text as M3Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.abracode.actionui.Common.ActionUIElement
import com.abracode.actionui.Common.ActionUIRegistry
import com.abracode.actionui.Common.ActionUIViewConstruction
import com.abracode.actionui.Common.LocalActionUILogger
import com.abracode.actionui.Common.buildChildModifier
import com.abracode.actionui.Helpers.BuildViewWithPopover
import com.abracode.actionui.Helpers.ProvideTextStyleEnvironment
import com.abracode.actionui.Helpers.stringProperty

/**
 * Pairs a `title` label with content. Mirror of the Apple `LabeledContent`
 * element (`ActionUI/Views/LabeledContent.swift`), which wraps
 * `SwiftUI.LabeledContent` - a leading label and trailing content that stays
 * visible in any container (unlike a control's built-in, often-hidden label).
 *
 * Rendered as a [Row]: the `title` leads, a flexible [Spacer] pushes the
 * `children` to the trailing edge (the SwiftUI label-left / content-right
 * layout). The cross axis is centered.
 *
 * **Supported properties.**
 *   * `title` - the leading label String (defaults to `""`); an empty title
 *     renders content only.
 *   * `children` - one or more trailing views; each gets the
 *     [androidx.compose.foundation.layout.RowScope] child modifier (so per-child
 *     `weight` / `align` work, as in [HStack]).
 *   * plus the universal modifiers resolved by `applyCommonProperties` (via
 *     [modifier]).
 */
object LabeledContent : ActionUIViewConstruction {
    @Composable
    override fun BuildView(element: ActionUIElement, modifier: Modifier) {
        val logger = LocalActionUILogger.current
        val title = element.properties?.stringProperty("title") ?: ""

        Row(
            modifier = modifier,
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.Start,
        ) {
            if (title.isNotEmpty()) {
                M3Text(title)
                Spacer(Modifier.width(8.dp))
            }
            Spacer(Modifier.weight(1f))
            element.children.orEmpty().forEach { child ->
                val builder = ActionUIRegistry.lookup(child.type) ?: return@forEach
                ProvideTextStyleEnvironment(child.properties, logger) {
                    builder.BuildViewWithPopover(child, buildChildModifier(child.properties, logger))
                }
            }
        }
    }
}

File: ActionUIWeb/src/Views/LabeledContent.js
// LabeledContent.js — LabeledContent element.
// Web analog of ActionUI/Views/LabeledContent.swift (and ActionUIAndroid
// Views/LabeledContent.kt).
//
// Pairs a `title` label with content: the title leads, the `children` are
// pushed to the trailing edge (the SwiftUI label-left / content-right layout,
// which stays visible in any container — unlike a control's built-in,
// often-hidden label). The row stretches to the available width like the
// Android weighted Row; an empty/absent title renders content only.
//
// Properties: title (String for the leading label, defaults to "").

import { register } from "../Common/ActionUIRegistry.js";
import { StackAxis } from "../Common/StackAxis.js";

register("LabeledContent", {
    valueType: "none",

    // Mirrors LabeledContent.swift validateProperties (warning text included verbatim).
    validateProperties: (properties, logger) => {
        const validated = { ...properties };
        if (validated.title !== undefined && typeof validated.title !== "string") {
            logger.log("LabeledContent 'title' must be String; setting to nil", "warning");
            delete validated.title;
        }
        return validated;
    },

    buildView: (element, properties, ctx) => {
        const node = document.createElement("div");
        node.className = "aui-stack aui-hstack aui-labeled-content";
        node.dataset.auiAxis = StackAxis.Horizontal;

        // The title flexes (flex:1 in theme.css), pushing the children to the
        // trailing edge; kept even when empty so content stays trailing.
        const title = document.createElement("span");
        title.className = "aui-labeled-content-title";
        title.textContent = properties.title ?? "";
        node.appendChild(title);

        for (const child of element.children()) {
            node.appendChild(ctx.build(child));
        }
        return node;
    },
});
